### ISSUE TITLE
Interaction detection is much slower than the original GA2M

### ISSUE THREAD

**Issue by MichaelDoron** [interpretml/interpret#126]
Hey,

I am comparing the results and computation time between the current python-based interpret EBM and the previous Java based GA2M, and the current implementation seems to take an order of magnitude more than the Java implementation (and the time increases with more interactions).

Am I doing something wrong? I thought this implementation is faster than the original GA2M.

To reproduce:
```

import sys
import os
import time
import numpy as np
from sklearn.metrics import average_precision_score
from interpret.glassbox import ExplainableBoostingRegressor

X_train = np.random.rand(1000, 3)
y_train = X_train[:, 0] * X_train[:,1]

file = open('data.attr', 'w')
for i in range(3):
  file.write('X_train{}: cont\n'.format(i))
file.write('label: cont (target)')
file.close()
np.savetxt('data.train', np.concatenate((X_train, y_train.reshape(-1,1)), axis=1), delimiter=' ')


dataset_file = 'data.train'
datasetDisc = 'data_disc.train'
attset = 'data.attr'
attsetDisc = 'data_disc.attr'
residualPath = 'data.res'
modelPath = 'model.model'
FASTOutput = 'result.out'

start = time.time()
os.system("/usr/bin/java mltk.core.processor.Discretizer -i {} -o {} -t {} -r {} -m {}".format(dataset_file, datasetDisc, dataset_file, attset, attsetDisc))
os.system("/usr/bin/java mltk.predictor.tree.ensemble.brt.LSBoostLearner -t {} -m {} -o {} -r {} -V False".format(datasetDisc, 1000, modelPath, attsetDisc))
os.system("/usr/bin/java mltk.predictor.evaluation.Predictor -d {} -m {} -r {} -R {}".format(datasetDisc, modelPath, attsetDisc, residualPath))
os.system("/usr/bin/java mltk.predictor.gam.interaction.FAST -d {} -m {} -r {} -R {} -o {} -b 8".format(dataset_file, modelPath, attset, residualPath, FASTOutput))
end = time.time()

print('Java GA2M took {} seconds'.format(end - start))


start = time.time()
ebm = ExplainableBoostingRegressor(interactions = 1, random_state=42)
ebm.fit(X_train, y_train)
ebm_global = ebm.explain_global()
end = time.time()

print('interpret EBM took {} seconds'.format(end - start))


```
The result is 

> Time: 1.744
RMSE on Test: 3.2977825941096555E-4
Reading residuals...
Discretizing attribute...
Generating all pairs of attributes...
Creating threads...
Running FAST...
Sorting pairs...
Time: 0.003
Java GA2M took 3.603398084640503 seconds
interpret EBM took 67.90893268585205 seconds

**Comment by interpret-ml** [comment]
Hi @MichaelDoron  --

A couple of things:

1) Interpret by default generates 16 outer bags and averages them afterwards, so to get an equivalent comparison, you'd need to run the java code 16 times.
2) mltk.predictor.gam.interaction.FAST just finds the pairs.  After running FAST you'd need to run mltk.predictor.gam.GA2MLearner to generate the GA2Ms, which is typically a lot slower than the FAST step.
3) Interpret runs the equivalent of mltk.predictor.gam.GA2MLearner twice.  The first run generates independent pairs for each of the outer bags.  The independent pairs are not guaranteed to be the same between outer bags, so there’s a step where they are combined to find the best overall pairs, then GA2MLearner is re-run on the global best pairs.

-- InterpretML team

**Comment by MichaelDoron** [comment]
Thank you for the fast reply! 

Regarding the second point, if I wanted to only score and rank the pairs, wouldn't just running mltk.predictor.gam.interaction.FAST suffice, without mltk.predictor.gam.GA2MLearner? And if it does suffice, is there an equivalent in interpret?

**Comment by interpret-ml** [comment]
@MichaelDoron --

You'll probably get better pair results if you use the Interpret implementation for pair detection.  You can get similar results using the java code, but you'd have to script some extra functionality not available in the released java code.  In Interpret, after running FAST, we examine the effect of both adding each pair to the original mains, and also the effect of removing each pair from the proposed set of pairs on a validation set.  Have a look at the _select_merged_pairs function if you want more details.  This is the function that would need to be scripted to get equivalent results using java.

The resulting pairs are exposed in inter_indices_.  Please note, this name will probably change in a near future release.

Here’s a pointer to the relevant section:
https://github.com/interpretml/interpret/blob/4d6942fd12f1e7c95236368f167690c476ad462f/python/interpret-core/interpret/glassbox/ebm/ebm.py#L1016

**Comment by paulbkoch** [comment]
As an update to the above, we now expose the pure FAST algorithm without the need to train an EBM.  More details in issue https://github.com/interpretml/interpret/issues/332
